L1Trigger/TrackFindingTracklet/interface/imath.h

```diff
@@ -0,0 +1,1137 @@
+          snprintf(slog, 100, "defining unitless constant, yet K is not a power of 2! %g, %g", K, pow(2, l));
+          edm::LogVerbatim("Tracklet") << slog;
+        }
+        Kmap_["2"] = l;
+      }
+    }
+    //construct from abother variable (all provenance info is lost!)
+    VarDef(imathGlobals *globals, std::string name, VarBase *p) : VarBase(globals, name, nullptr, nullptr, nullptr, 1) {
+      op_ = "def";
+      K_ = p->K();
+      nbits_ = p->nbits();
+      Kmap_ = p->Kmap();
+    }
+    void set_fval(double fval) {
+      fval_ = fval;
+      if (fval > 0)
+        ival_ = fval / K_;
+      else
+        ival_ = fval / K_ - 1;
+      val_ = ival_ * K_;
+    }
+    void set_ival(int ival) {
+      ival_ = ival;
+      fval_ = ival * K_;
+      val_ = ival_ * K_;
+    }
+    ~VarDef() override = default;
+    void print(std::ofstream &fs, Verilog, int l1 = 0, int l2 = 0, int l3 = 0) override;
+    void print(std::ofstream &fs, HLS, int l1 = 0, int l2 = 0, int l3 = 0) override;
+  };
+
+  class VarAdd : public VarBase {
+  public:
+    VarAdd(imathGlobals *globals, std::string name, VarBase *p1, VarBase *p2, double range = -1, int nmax = 18)
+        : VarBase(globals, name, p1, p2, nullptr, 1) {
+      op_ = "add";
+
+      std::map<std::string, int> map1 = p1->Kmap();
+      std::map<std::string, int> map2 = p2->Kmap();
+      int s1 = map1["2"];
+      int s2 = map2["2"];
+
+      //first check if the constants are all lined up
+      //go over the two maps subtracting the units
+      for (const auto &it : map2) {
+        if (map1.find(it.first) == map1.end())
+          map1[it.first] = -it.second;
+        else
+          map1[it.first] = map1[it.first] - it.second;
+      }
+
+      char slog[100];
+
+      //assert if different
+      for (const auto &it : map1) {
+        if (it.second != 0) {
+          if (it.first != "2") {
+            snprintf(
+                slog, 100, "VarAdd: bad units! %s^%i for variable %s", (it.first).c_str(), it.second, name_.c_str());
+            edm::LogVerbatim("Tracklet") << slog;
+            edm::LogVerbatim("Tracklet") << " *********************************************************";
+            p1->dump_msg();
+            edm::LogVerbatim("Tracklet") << " *********************************************************";
+            p2->dump_msg();
+            assert(0);
+          }
+        }
+      }
+
+      double ki1 = p1->K() / pow(2, s1);
+      double ki2 = p2->K() / pow(2, s2);
+      //those should be the same
+      if (std::abs(ki1 / ki2 - 1.) > 1e-6) {
+        snprintf(slog, 100, "VarAdd: bad constants! %f %f for variable %s", ki1, ki2, name_.c_str());
+        edm::LogVerbatim("Tracklet") << slog;
+        edm::LogVerbatim("Tracklet") << " *********************************************************";
+        p1->dump_msg();
+        edm::LogVerbatim("Tracklet") << " *********************************************************";
+        p2->dump_msg();
+        assert(0);
```

Comment:
please avoid unnecessary excess text in messages, and use exception rather than assert(0)

Reply:
(everywhere in this file)

Reply:
fixed